File: main/java/com/marsrovers/location/Plateau.java
package com.marsrovers.location;

import com.marsrovers.units.Rover;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Plateau {

    final int horizontalBoundary;
    final int verticalBoundary;
    final Map<String, Rover> rovers = new HashMap<>();

    Plateau(int horizontalBoundary, int verticalBoundary) {

        if (horizontalBoundary <= 0) {
            throw new IllegalArgumentException("Horizontal Boundary must be grater than 0");
        }

        if (verticalBoundary <= 0) {
            throw new IllegalArgumentException("Vertical Boundary must be grater than 0");
        }

        this.horizontalBoundary = horizontalBoundary;
        this.verticalBoundary = verticalBoundary;
    }

    public boolean isOccupied(final Rover rover) {
        return rovers
                .entrySet()
                .stream()
                .anyMatch(es-> es.getValue().position.equals(rover.position));
    }

    public void set(Rover rover) {
        rovers.put(rover.id, rover);
    }

    public Optional<Rover> get(String roverId) {
        return Optional.ofNullable(rovers.get(roverId));
    }


    @Override
    public String toString() {
        final StringBuilder builder = new StringBuilder();

        rovers.entrySet().forEach(es -> builder.append("[").append(es.getValue().toString()).append("]"));

        return builder.toString();

    }
}

File: main/java/com/marsrovers/location/Plateau.kt
package com.marsrovers.location

import com.marsrovers.units.Rover

import java.util.HashMap
import java.util.Optional

class Plateau internal constructor(val horizontalBoundary: Int, val verticalBoundary: Int) {
    internal val rovers: MutableMap<String, Rover> = HashMap()

    init {

        if (horizontalBoundary <= 0) {
            throw IllegalArgumentException("Horizontal Boundary must be grater than 0")
        }

        if (verticalBoundary <= 0) {
            throw IllegalArgumentException("Vertical Boundary must be grater than 0")
        }
    }

    fun isOccupied(rover: Rover): Boolean = rovers.entries.any { es -> es.value.position == rover.position }

    fun set(rover: Rover) = rovers.put(rover.id, rover)

    operator fun get(roverId: String): Optional<Rover> = Optional.ofNullable<Rover>(rovers[roverId])

    override fun toString(): String =  rovers.entries.joinToString(prefix = "[", postfix = "]")

}
